### PR TITLE
Added support for QOI format

### PR DESCRIPTION
**Pull request by jmaselbas**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/ImageMagick/ImageMagick/pulls) open
- [x] I have verified that I am following the existing coding patterns and practices as demonstrated in the repository.

### Description
Add support for QOI image format.
Reference implementation can be found at https://github.com/phoboslab/qoi ;
The QOI image file format specification can be found at:  https://qoiformat.org/qoi-specification.pdf
This is not an original implementation, it is mostly an adaptation of the reference implementation for ImageMagick

cc @phoboslab

### CODE REVIEW

**Comment by urban-warrior** [comment]
This command fails: `convert logo: logo.qoi`.  You can handle colormapped pixels or you could promote the PseudoClass storage format to DirectClass.

These commands fail:

```
$ magick logo: -type truecolor logo.qoi
$ magick logo.qoi logo.pnm
magick: unexpected end-of-file 'logo.qoi': Invalid argument @ error/qoi.c/ReadQOIImage/266.
```

Can you add a patch to resolve this issue?  Also, `splint` returns these warnings, can you correct these issues?

```
coders/qoi.c: (in function ReadQOIImage)
coders/qoi.c:158:3: Assignment of int to size_t: channels = ReadBlobByte(image)
  To allow arbitrary integral types to match long unsigned, use +longintegral.
coders/qoi.c:166:3: Assignment of int to size_t:
                       colorspace = ReadBlobByte(image)
coders/qoi.c:168:5: Return value (type MagickBooleanType) ignored:
                       SetImageColorspa...
  Result returned by function call is not used. If this is intended, can cast
  result to (void) to eliminate message. (Use -retvalother to inhibit warning)
coders/qoi.c:170:5: Return value (type MagickBooleanType) ignored:
                       SetImageColorspa...
coders/qoi.c:201:13: Operands of < have incompatible types (ssize_t, arbitrary
    unsigned integral type): p < (image->rows * image->columns)
  To allow arbitrary integral types to match any integral type, use
  +matchanyintegral.
coders/qoi.c:210:7: Assignment of int to unsigned char: px.rgba.r = b
  To ignore signs in type comparisons use +ignoresigns
coders/qoi.c:213:7: Assignment of int to unsigned char: px.rgba.g = b
coders/qoi.c:216:7: Assignment of int to unsigned char: px.rgba.b = b
coders/qoi.c:221:7: Assignment of int to unsigned char: px.rgba.r = b
coders/qoi.c:224:7: Assignment of int to unsigned char: px.rgba.g = b
coders/qoi.c:227:7: Assignment of int to unsigned char: px.rgba.b = b
coders/qoi.c:230:7: Assignment of int to unsigned char: px.rgba.a = b
coders/qoi.c:261:48: Function SetImageProgress expects arg 3 to be
                        MagickOffsetType gets ssize_t: p
coders/qoi.c:271:7: Return value (type MagickBooleanType) ignored:
                       SyncImage(image,...
coders/qoi.c: (in function WriteQOIImage)
coders/qoi.c:349:30: Function WriteBlobByte expects arg 2 to be unsigned char
                        gets size_t: channels
coders/qoi.c:350:30: Function WriteBlobByte expects arg 2 to be unsigned char
                        gets size_t: colorspace
coders/qoi.c:379:36: Function WriteBlobByte expects arg 2 to be unsigned char
                        gets arbitrary unsigned integral type: 0xc0 | (run - 1)
  To ignore type qualifiers in type comparisons use +ignorequals.
coders/qoi.c:385:34: Function WriteBlobByte expects arg 2 to be unsigned char
                        gets arbitrary unsigned integral type: 0xc0 | (run - 1)
coders/qoi.c:401:7: Assignment of unsigned char to char:
                       vr = px.rgba.r - pp.rgba.r
coders/qoi.c:402:7: Assignment of unsigned char to char:
                       vg = px.rgba.g - pp.rgba.g
coders/qoi.c:403:7: Assignment of unsigned char to char:
                       vb = px.rgba.b - pp.rgba.b
coders/qoi.c:410:36: Function WriteBlobByte expects arg 2 to be unsigned char
    gets int: 0x40 | (vr + 2) << 4 | (vg + 2) << 2 | (vb + 2)
coders/qoi.c:415:36: Function WriteBlobByte expects arg 2 to be unsigned char
                        gets int: 0x80 | (vg + 32)
coders/qoi.c:416:36: Function WriteBlobByte expects arg 2 to be unsigned char
                        gets int: (vg_r + 8) << 4 | (vg_b + 8)
```

**Comment by jmaselbas** [comment]
Hi, thanks for the review, I will try to take a look at theses issues and warnings.
I must admit that I don't really understand the current code base and most of the names used here are obscure to me.
So I have no idea how to handle "colormapped pixels" nor what's the meaning behind PseudoClass vs DirectClass.

**Comment by urban-warrior** [comment]
Do what you can and we'll add a few patches.  If the coder is not robust by the next release, perhaps a few weeks from now, we may not be able to keep it in the release.

File: coders/qoi.c [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+  if (quantum_info == (QuantumInfo *) NULL)
+    ThrowWriterException(ImageError,"MemoryAllocationFailed");
+
+  quantum_type=GetQuantumType(image,exception);
+  if (quantum_type == RGBQuantum)
+    channels=3;
+  else if (quantum_type == RGBAQuantum)
+    channels=4;
+  else
+    ThrowWriterException(ImageError,"MemoryAllocationFailed");
```
Comment (phoboslab) [inline coders/qoi.c:335]: Is that the right exception message to throw here?

File: coders/qoi.c [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+        (void) WriteBlobByte(image,px.rgba.b);
+      }
+    } else {
+      (void) WriteBlobByte(image,QOI_OP_RGBA);
+      (void) WriteBlobByte(image,px.rgba.r);
+      (void) WriteBlobByte(image,px.rgba.g);
+      (void) WriteBlobByte(image,px.rgba.b);
+      (void) WriteBlobByte(image,px.rgba.a);
+    }
+  }
```
Comment (phoboslab) [inline coders/qoi.c:430]: The 8-Byte end-marker (`0x0000000000000001`) is not written.

File: coders/qoi.c [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+      vg_b=vb - vg;
+
+      if (vr > -3 && vr < 2 &&
+          vg > -3 && vg < 2 &&
+          vb > -3 && vb < 2) {
+        (void) WriteBlobByte(image,QOI_OP_DIFF
+                 | (vr + 2) << 4 | (vg + 2) << 2 | (vb + 2));
+      } else if (vg_r >  -9 && vg_r < 8 &&
+                 vg   > -33 && vg   < 32 &&
+                 vg_b >  -9 && vg_b < 2) {
```
Comment (phoboslab) [inline coders/qoi.c:414]: Should be `vg_b < 8`

File: coders/qoi.c (hunk 1 above)
Comment (urban-warrior) [reply coders/qoi.c:335]: Try `ThrowWriterException(CoderError,"ImageTypeNotSupported");`

**Comment by jmaselbas** [comment]
@phoboslab Thanks for the review I will send a fixup asap

**Comment by jmaselbas** [comment]
@urban-warrior Could you provide the `splint` command line you've used, It would help me to fix the warnings, thanks.

**Comment by dlemstra** [comment]
@jmaselbas I just pushed some minor patches to the `qoi.c` file that hopefully does not break your PR because it mostly are comments.

File: coders/qoi.c [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+
+  status=SetImageExtent(image,image->columns,image->rows,exception);
+  if (status == MagickFalse)
+    return(DestroyImageList(image));
+  quantum_info=AcquireQuantumInfo(image_info,image);
+  if (quantum_info == (QuantumInfo *) NULL)
+    ThrowReaderException(ResourceLimitError,"MemoryAllocationFailed");
+  status=SetQuantumFormat(image,quantum_info,UnsignedQuantumFormat);
+  if (status == MagickFalse)
+    return(DestroyImageList(image));
```
Comment (dlemstra) [inline coders/qoi.c:182]: Leaking `quantum_info` here and you don't appear to be using it so we better remove this variable?

File: coders/qoi.c [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+    status=SetImageProgress(image,LoadImageTag,p,image->rows * image->columns);
+    if (status == MagickFalse)
+      break;
+  }
+  if (EOFBlob(image) != MagickFalse)
+    ThrowFileException(exception,CorruptImageError,
+      "UnexpectedEndOfFile",image->filename);
+  if (status != MagickFalse) {
+    status=SyncAuthenticPixels(image,exception);
+    if (image->ping == MagickFalse)
```
Comment (dlemstra) [inline coders/qoi.c:270]: We can do an ealier exit when we only are pinging the image?

File: coders/qoi.c [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+  return(GetFirstImageInList(image));
+}
+

+
+static MagickBooleanType WriteQOIImage(const ImageInfo *image_info,Image *image,
+  ExceptionInfo *exception)
+{
+  QuantumInfo
+    *quantum_info;
```
Comment (dlemstra) [inline coders/qoi.c:284]: This also seems to be unused?

File: coders/qoi.c [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+  quantum_type=GetQuantumType(image,exception);
+  if (quantum_type == RGBQuantum)
+    channels=3;
+  else if (quantum_type == RGBAQuantum)
+    channels=4;
+  else
+    ThrowWriterException(ImageError,"MemoryAllocationFailed");
+
+  if (IssRGBCompatibleColorspace(image->colorspace) == MagickFalse)
+   (void) TransformImageColorspace(image,sRGBColorspace,exception);
```
Comment (dlemstra) [inline coders/qoi.c:338]: Seems to be missing a space here but don't worry about that. We will do a style cleanup when you are done.

File: coders/qoi.c [hunk 8]
```diff
@@ -0,0 +1,450 @@
...
+  */
+  px.rgba.r=0;
+  px.rgba.g=0;
+  px.rgba.b=0;
+  px.rgba.a=255;
+  memset(lut,0,sizeof(lut));
+  run=0;
+  p=GetVirtualPixels(image,0,0,image->columns,image->rows,exception);
+  if (p == (const Quantum *) NULL)
+    return(MagickFalse);
```
Comment (dlemstra) [inline coders/qoi.c:362]: Better to do ` ThrowWriterException(ResourceLimitError,"MemoryAllocationFailed");` here instead.

File: coders/qoi.c [hunk 9]
```diff
@@ -0,0 +1,450 @@
...
+      }
+    } else {
+      (void) WriteBlobByte(image,QOI_OP_RGBA);
+      (void) WriteBlobByte(image,px.rgba.r);
+      (void) WriteBlobByte(image,px.rgba.g);
+      (void) WriteBlobByte(image,px.rgba.b);
+      (void) WriteBlobByte(image,px.rgba.a);
+    }
+  }
+  return(MagickTrue);
```
Comment (dlemstra) [inline coders/qoi.c:431]: A call to `(void) CloseBlob(image);` is missing before the return and this will keep the file.

**Review by dlemstra — COMMENTED** [review]
Could you please also adress these comments in your follow up PR?

And thanks for helping us add support for this format.

File: coders/qoi.c (hunk 5 above)
Comment (jmaselbas) [reply coders/qoi.c:270]: I didn't understood what ping mean, I've copied from other coders... I don't understand what you are suggesting sorry.

File: coders/qoi.c (hunk 7 above)
Comment (jmaselbas) [reply coders/qoi.c:338]: done

File: coders/qoi.c (hunk 8 above)
Comment (jmaselbas) [reply coders/qoi.c:362]: done

File: coders/qoi.c (hunk 9 above)
Comment (jmaselbas) [reply coders/qoi.c:431]: done

File: coders/qoi.c (hunk 5 above)
Comment (dlemstra) [reply coders/qoi.c:270]: `ping` means that we only want the metadata of the image. So you can do something like this before you read the pixels:

```
if (image->ping != MagickFalse)
  return(GetFirstImageInList(image))
```

**Pull request merged**